hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlPlanCacheTest.java

```diff
@@ -123,6 +128,24 @@ public void test_dmlCaching() {
         assertThat(planCache(instance()).size()).isZero();
     }
 
+    @Test
+    @Ignore // TODO: [sasha] enable after IMDG engine removal
+    public void test_index() {
+        IMap<Object, Object> map = instance().getMap("m");
+
+        createMapping("map", map.getName(), "id", PersonId.class, "varchar");
+        String indexName = randomName();
+
+        map.addIndex(new IndexConfig(IndexType.SORTED, "__key.id").setName(indexName));
+        sqlService.execute("SELECT * FROM map ORDER BY id");
+        assertThat(planCache(instance()).size()).isEqualTo(1);
+
+        getMapContainer(map).getIndexes().destroyIndexes();
+        map.addIndex(new IndexConfig(IndexType.HASH, "__key.id").setName(indexName));
+
+        assertTrueEventually(() -> assertThat(planCache(instance()).size()).isZero());
```

Comment:
If I understand correctly, we rely on periodic check to invalidate plans with indexes. So far it wasn't needed for explicit mappings. What happens if I remove index and execute query before plan invalidation?

Reply:
> If I understand correctly, we rely on periodic check to invalidate plans with indexes.

You are right. 

First of all, index removal is a private API. IMHO, the more probabilistic scenario could be index creation with same name after map destroy. Of course, index removal still can be used, but in such case I don't expect that the window between index removal and map querying would be so small (in terms of time)
Second, this issue and similar issues are known and are intended to fix in 5.1 (e.g. #19130)